File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        count = 0
        m = len(grid)
        n = len(grid[0])

        # find the number of fresh fruits and the position of the rotten fruits
        fresh = 0
        rotten = deque()
        for y in range(m):
            for x in range(n):
                if grid[y][x]==2:
                    rotten.append([y, x])
                elif grid[y][x]==1:
                    fresh+=1
        
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]
        # while there are still fresh fruit remain
        while fresh!=0:
            count += 1
            # to check whether there are fresh fruit that can still be changed
            changed = False
            # bfs
            for x in range(len(rotten)):
                pos = rotten.popleft()
                # check if fresh fruit exist in the four directions of the rotten fruit
                for y in range(4):
                    if dy[y]+pos[0]<m and dy[y]+pos[0]>-1 and dx[y]+pos[1]<n and dx[y]+pos[1]>-1:
                        if grid[dy[y]+pos[0]][dx[y]+pos[1]]==1:
                            fresh-=1
                            grid[dy[y]+pos[0]][dx[y]+pos[1]] = 2
                            rotten.append([dy[y]+pos[0], dx[y]+pos[1]])
                            changed = True
            # if no fruit is changed (this guarentees fresh fruit remains)
            if not changed:
                return -1
        return count

File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int ans = 0;
        int ROW = grid.size(), COL = grid[0].size();
        int dr[4] = {-1, 1, 0, 0};
        int dc[4] = {0, 0, -1, 1};

        // Use queue to store all rotten orange
        queue<vector<int>> q;
        // Find all exist rotten orange
        for (int r=0; r<ROW; ++r) {
            for (int c=0; c<COL; ++c) {
                if (grid[r][c] == 2)
                    q.push({r, c});
            }
        }

        // Use BFS to iterate through all oranges that can be infected
        while (!q.empty()) {
            // iterate this level
            int n = q.size();
            for (n; n>0; --n) {
                vector<int> curr = q.front();
                q.pop();

                // Check if there is fresh orange in its neighbors
                int nxt_r, nxt_c;
                for (int i=0; i<4; ++i) {
                    nxt_r = curr[0] + dr[i];
                    nxt_c = curr[1] + dc[i];

                    // first check if out of bound, then check if it is fresh orange
                    if (0 <= nxt_r && nxt_r < ROW && 0 <= nxt_c && nxt_c < COL && grid[nxt_r][nxt_c]==1) {
                        // push it into the queue for bfs
                        q.push({nxt_r, nxt_c});
                        // mark it as rotten
                        grid[nxt_r][nxt_c] = 2;
                    }
                }
            }

            ans++;
        }

        for (int r=0; r<ROW; ++r) {
            for (int c=0; c<COL; ++c) {
                if (grid[r][c] == 1)
                    return -1;
            }
        }

        return (ans > 0) ? ans - 1 : ans;
    }
};